core/state/statedb.go

```diff
@@ -900,15 +900,28 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 		}
 	}
 	usedAddrs := make([][]byte, 0, len(s.stateObjectsPending))
+	// Perform updates before deletions.  This prevents resolution of unnecessary trie nodes
+	// in circumstances similar to the following:
+	//
+	// Take value nodes 1, 2 who share the same full node parent 3 and have no other siblings.
+	// 1 self-destructs specifying a non-existing recipient account which would be a child of 3.
+	// If the deletion of 1 happens before the account-creating balance transfer, 3 will temporarily
+	// be collapsed to a short node on 2, requiring 2 to be unecessarily resolved.
```

Comment:
```suggestion
	// Consider nodes `A` and `B` who share the same full node parent `P` and have no other siblings.
	// During the execution of a block:
	// - `A` self-destructs, 
	// - `C` is created, and also shares the parent `P`. 
	// If the self-destruct is handled first, then `P` would be left with only one child, thus collapsed
	// into a shortnode. This requires `B` to be resolved from disk. 
	// Whereas if the created node is handled first, then the collapse is avoided, and `B` is not resolved.
```
IMO a bit clearer? Also I don't see what the beneficiary matters, really. 

Secondly, shouldn't there be a corresponding change in storage? At line `301` in `state_object.go`: 
```golang
	for key, value := range s.pendingStorage {
...
		if (value == common.Hash{}) {
			if err := tr.DeleteStorage(s.address, key[:]); err != nil {
...
		} else {
...
			if err := tr.UpdateStorage(s.address, key[:], trimmed); err != nil {
```

Reply:
Right.  And also worth noting that while the account case was caught by the state test suite, the storage case is not and could use a test case (TODO: me).